sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala

```diff
@@ -128,11 +130,39 @@ class QueryExecution(
     case _ => "command"
   }
 
+  private def assertNoStreamSourceMarkerNode(p: LogicalPlan): Unit = {
+    // In UnsupportedOperationChecker.checkForBatch, we check if the plan has any streaming node.
+    // That is more aggressive than just checking the marker node for streaming source which is
+    // yet to be materialized. We'd like to be a bit conservative here since this is the exact
+    // problematic case we figured out.
+    p.foreach {
+      case _: StreamingRelation | _: StreamingRelationV2 |
+           _: StreamingExecutionRelation | _: StreamingDataSourceV2ScanRelation =>
+        val msg = "Queries with streaming sources must be executed with writeStream.start()"
```

Comment:
Does it have to be `writeStream`? Can it be `readStream`?

Reply:
Oh, `readStream` cannot start a streaming query.

Reply:
Yeah, for now, every streaming query should be triggered via `DataStreamWriter.start()`.